src/types.ts

```diff
@@ -4,20 +4,40 @@
  */
 
 /* eslint-disable @typescript-eslint/no-empty-interface */
-
-import type * as express from 'express';
 import type * as http from 'http';
 import type * as httpProxy from 'http-proxy';
 import type * as net from 'net';
 import type * as url from 'url';
 
-export interface Request extends express.Request {}
-export interface Response extends express.Response {}
+export type Request = http.IncomingMessage;
+export type Response = http.ServerResponse;
 
-export interface RequestHandler extends express.RequestHandler {
-  upgrade?: (req: Request, socket: net.Socket, head: any) => void;
+/**
+ * http-proxy-middleware supports framework specific values. The following
+ * values are primarily decorated onto IncomingMessage by express, but are
+ * not required for use.
+ */
+declare module 'http' {
+  interface IncomingMessage {
+    originalUrl?: string;
+    hostname?: string;
+    host?: string;
+    body?: Record<string, any>;
+  }
 }
 
+export type Next = (...args: unknown[]) => unknown;
+
+type RequestMiddlewareFunction = (
+  req: http.IncomingMessage,
+  res: http.ServerResponse,
+  next: Next
+) => unknown | Promise<unknown>;
```

Comment:
Given that this function is using a callback, should it instead be `=> void | Promise<void>`?

Reply:
we could, but I'd suggest that such a type may be less portable it would force all middleware providers to actually be void. i think they generally _are_ void (connect, express, koa), which makes such a suggestion quite reasonable. however, `unknown` doesn't hurt--as those middleware consumers aren't ever going to consume those values. for middlewares that do return a value, using `void` means they wouldn't be able to use `http-proxy-middleware` typed, and i'd like to err on the side of supporting any such users

Reply:
in other worrds, if `marcs-cool-server` used a middleware type of `(req,res,next) => res`, we'd be in trouble. at the office, i have some cohorts who were _quite passionate_ about res returning middleware, which i didn't necessarily disagree with. but know that such impls do exist out there in userspace. perhaps not by the big players.